internal/mapdecode/hooks.go

```diff
@@ -175,3 +185,124 @@ func strconvHook(from, to reflect.Type, data reflect.Value) (reflect.Value, erro
 
 	return data, nil
 }
+
+// fieldHook applies the user-specified FieldHookFunc to all struct fields.
+func fieldHook(opts *options) reflectHook {
+	hook := opts.FieldHook
+	return func(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+		if to.Kind() != reflect.Struct || from.Kind() != reflect.Map {
+			return data, nil
+		}
+
+		// We can only decode map[string]* and map[interface{}]* into structs.
+		if k := from.Key().Kind(); k != reflect.String && k != reflect.Interface {
+			return data, nil
+		}
+
+		// This map tracks type-changing updates to items in the map.
+		//
+		// If the source map has a rigid type for values (map[string]string
+		// rather than map[string]interface{}), we can't make replacements to
+		// values in-place if a hook changed the type of a value. So we will
+		// make a copy of the source map with a more liberal type and inject
+		// these updates into the copy.
+		updates := make(map[interface{}]interface{})
+
+		var errors []error
+		for i := 0; i < to.NumField(); i++ {
+			structField := to.Field(i)
+			if structField.PkgPath != "" {
```

Comment:
not sure if it applies here, but take a look at:
https://github.com/golang/go/issues/12367#issuecomment-151142342